### PR TITLE
fix: Make null key serialization/deserialization symmetrical

### PR DESCRIPTION
**Pull request by purplefox**
### Description 

Fixes https://github.com/confluentinc/ksql/issues/4109

This fix makes sure that null keys are preserved during serialization followed by deserialization. I.e. the serialization/deserialization for null is now symmetrical.

Previously if a null key was serialized then deserialized it would end up as Struct{} with one null element, not the value null. 

This meant that Streams graceful null key checking didn't work as the key wasn't null (it was Struct{}) as it was passed through the streams topology. When an attempt to extract the actual key before inserting into RocksDB occurred it gave an NPE.

Now we actually pass null as the key for the ConsumerRecord that is passed through the topology. This allows the streams null key checking to do it's job and for the null key record to be handled gracefully.

### Testing done 

Added new QTT test

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-serde/src/main/java/io/confluent/ksql/serde/kafka/KafkaSerdeFactory.java [hunk 1]
```diff
@@ -142,7 +142,7 @@ public Struct deserialize(final String topic, final byte[] bytes) {
         final Object primitive = delegate.deserialize(topic, bytes);
         final Struct struct = new Struct(schema);
         struct.put(field, primitive);
-        return struct;
+        return primitive == null ? null : struct;
```
Comment (big-andy-coates) [inline ksql-serde/src/main/java/io/confluent/ksql/serde/kafka/KafkaSerdeFactory.java]: nit: if `primitive` is `null`, then could return after line 142.

File: ksql-serde/src/test/java/io/confluent/ksql/serde/kafka/KafkaSerdeFactoryTest.java [hunk 2]
```diff
@@ -192,7 +194,11 @@ private void shouldHandle(final SqlType fieldSchema, final Object value) {
     final Object result = serde.deserializer().deserialize("topic", bytes);
 
     // Then:
-    assertThat(result, is(struct));
+    if (value == null) {
+      assertThat(result, is(nullValue()));
+    } else {
+      assertThat(result, is(struct));
+    }
```
Comment (big-andy-coates) [inline ksql-serde/src/test/java/io/confluent/ksql/serde/kafka/KafkaSerdeFactoryTest.java]: if its null, then assert its null... weird asserting! ;)

Also, this means all of the tests will pass if deserialize always returns `null`.

I think it would be more explicit if we revert this change and instead have explicit tests that test we:

a) serialize `null` as `null` , and
b) deserialize `null` as `null`.

Just played around with this, so may as well share the code.

i.e. we revert this to:

```java
  assertThat(result, is(struct));
```

And replace `shouldHandleNulls` with:

```java
@Test
  public void shouldSerializeNullAsNull() {
    // Given:
    final PersistenceSchema schema = schemaWithFieldOfType(SqlTypes.INTEGER);

    final Serde<Object> serde = factory.createSerde(schema, ksqlConfig, srClientFactory);

    // When:
    final byte[] result = serde.serializer().serialize("topic", null);

    // Then:
    assertThat(result, is(nullValue()));
  }

  @Test
  public void shouldDeserializeNullAsNull() {
    // Given:
    final PersistenceSchema schema = schemaWithFieldOfType(SqlTypes.INTEGER);

    final Serde<Object> serde = factory.createSerde(schema, ksqlConfig, srClientFactory);

    // When:
    final Object result = serde.deserializer().deserialize("topic", null);

    // Then:
    assertThat(result, is(nullValue()));
  }
```
